File: backend/src/utils/safeZones.ts
export interface BBox {
  top: number;    // 0-1000 normalized
  left: number;
  width: number;
  height: number;
}

export interface SafeZone extends BBox {
  area: number;
  label: string;  // e.g. "top-left", "bottom-center"
}

const MIN_ZONE_WIDTH = 150;
const MIN_ZONE_HEIGHT = 50;
const CANVAS_SIZE = 1000;

/**
 * Subtract a single obstacle bbox from a list of available rectangles.
 * Splits each affected rectangle into up to 4 sub-rects around the obstacle.
 */
function subtractBBox(available: BBox[], obstacle: BBox): BBox[] {
  if (obstacle.width <= 0 || obstacle.height <= 0) return [...available];
  const result: BBox[] = [];
  for (const rect of available) {
    const rRight = rect.left + rect.width;
    const rBottom = rect.top + rect.height;
    const oRight = obstacle.left + obstacle.width;
    const oBottom = obstacle.top + obstacle.height;

    // No overlap — keep rect as-is
    if (
      rRight <= obstacle.left ||
      rect.left >= oRight ||
      rBottom <= obstacle.top ||
      rect.top >= oBottom
    ) {
      result.push(rect);
      continue;
    }

    // Top strip (above obstacle)
    if (rect.top < obstacle.top) {
      result.push({
        top: rect.top,
        left: rect.left,
        width: rect.width,
        height: obstacle.top - rect.top,
      });
    }
    // Bottom strip (below obstacle)
    if (rBottom > oBottom) {
      result.push({
        top: oBottom,
        left: rect.left,
        width: rect.width,
        height: rBottom - oBottom,
      });
    }
    // Left strip (left of obstacle, bounded vertically by obstacle)
    if (rect.left < obstacle.left) {
      result.push({
        top: Math.max(rect.top, obstacle.top),
        left: rect.left,
        width: obstacle.left - rect.left,
        height: Math.min(rBottom, oBottom) - Math.max(rect.top, obstacle.top),
      });
    }
    // Right strip (right of obstacle, bounded vertically by obstacle)
    if (rRight > oRight) {
      result.push({
        top: Math.max(rect.top, obstacle.top),
        left: oRight,
        width: rRight - oRight,
        height: Math.min(rBottom, oBottom) - Math.max(rect.top, obstacle.top),
      });
    }
  }
  return result;
}

function zoneLabel(zone: BBox): string {
  const centerX = zone.left + zone.width / 2;
  const centerY = zone.top + zone.height / 2;
  const vert = centerY < 333 ? "top" : centerY > 666 ? "bottom" : "center";
  const horiz = centerX < 333 ? "left" : centerX > 666 ? "right" : "center";
  return vert === horiz ? vert : `${vert}-${horiz}`;
}

/**
 * Compute safe zones for text placement given a list of obstacle bboxes.
 * Returns zones sorted by area descending, filtered to minimum viable size.
 * Each obstacle gets 20px padding to avoid text sitting right on the edge.
 */
export function computeSafeZones(obstacles: BBox[]): SafeZone[] {
  let available: BBox[] = [
    { top: 0, left: 0, width: CANVAS_SIZE, height: CANVAS_SIZE },
  ];

  for (const obstacle of obstacles) {
    // Add safety padding around each obstacle
    const paddedLeft   = Math.max(0, obstacle.left - 20);
    const paddedTop    = Math.max(0, obstacle.top - 20);
    const paddedRight  = Math.min(CANVAS_SIZE, obstacle.left + obstacle.width + 20);
    const paddedBottom = Math.min(CANVAS_SIZE, obstacle.top + obstacle.height + 20);
    const padded: BBox = {
      top:    paddedTop,
      left:   paddedLeft,
      width:  paddedRight  - paddedLeft,
      height: paddedBottom - paddedTop,
    };
    available = subtractBBox(available, padded);
  }

  return available
    .filter((r) => r.width >= MIN_ZONE_WIDTH && r.height >= MIN_ZONE_HEIGHT)
    .map((r) => ({
      ...r,
      area: r.width * r.height,
      label: zoneLabel(r),
    }))
    .sort((a, b) => b.area - a.area);
}

File: backend-python/app/utils/safe_zones.py
from __future__ import annotations

from dataclasses import dataclass

MIN_ZONE_WIDTH = 150
MIN_ZONE_HEIGHT = 50
CANVAS_SIZE = 1000
MARGIN = 30


@dataclass
class BBox:
    top: float
    left: float
    width: float
    height: float


@dataclass
class SafeZone(BBox):
    area: float = 0
    label: str = ""


@dataclass
class TextSuggestion:
    part: str = ""
    preferred_zone: str | None = None
    style: dict | None = None
    position: BBox | None = None


def _subtract_bbox(available: list[BBox], obstacle: BBox) -> list[BBox]:
    if obstacle.width <= 0 or obstacle.height <= 0:
        return list(available)

    result: list[BBox] = []
    for rect in available:
        r_right = rect.left + rect.width
        r_bottom = rect.top + rect.height
        o_right = obstacle.left + obstacle.width
        o_bottom = obstacle.top + obstacle.height

        if r_right <= obstacle.left or rect.left >= o_right or r_bottom <= obstacle.top or rect.top >= o_bottom:
            result.append(rect)
            continue

        # Top strip
        if rect.top < obstacle.top:
            result.append(BBox(top=rect.top, left=rect.left, width=rect.width, height=obstacle.top - rect.top))
        # Bottom strip
        if r_bottom > o_bottom:
            result.append(BBox(top=o_bottom, left=rect.left, width=rect.width, height=r_bottom - o_bottom))
        # Left strip
        if rect.left < obstacle.left:
            clip_top = max(rect.top, obstacle.top)
            clip_bottom = min(r_bottom, o_bottom)
            result.append(BBox(top=clip_top, left=rect.left, width=obstacle.left - rect.left, height=clip_bottom - clip_top))
        # Right strip
        if r_right > o_right:
            clip_top = max(rect.top, obstacle.top)
            clip_bottom = min(r_bottom, o_bottom)
            result.append(BBox(top=clip_top, left=o_right, width=r_right - o_right, height=clip_bottom - clip_top))

    return result


def _zone_label(zone: BBox) -> str:
    center_x = zone.left + zone.width / 2
    center_y = zone.top + zone.height / 2
    vert = "top" if center_y < 333 else ("bottom" if center_y > 666 else "center")
    horiz = "left" if center_x < 333 else ("right" if center_x > 666 else "center")
    return vert if vert == horiz else f"{vert}-{horiz}"


def compute_safe_zones(obstacles: list[BBox]) -> list[SafeZone]:
    available: list[BBox] = [BBox(top=0, left=0, width=CANVAS_SIZE, height=CANVAS_SIZE)]

    for obstacle in obstacles:
        padded_left = max(0, obstacle.left - 20)
        padded_top = max(0, obstacle.top - 20)
        padded_right = min(CANVAS_SIZE, obstacle.left + obstacle.width + 20)
        padded_bottom = min(CANVAS_SIZE, obstacle.top + obstacle.height + 20)
        padded = BBox(top=padded_top, left=padded_left, width=padded_right - padded_left, height=padded_bottom - padded_top)
        available = _subtract_bbox(available, padded)

    zones = [
        SafeZone(
            top=r.top, left=r.left, width=r.width, height=r.height,
            area=r.width * r.height, label=_zone_label(r),
        )
        for r in available
        if r.width >= MIN_ZONE_WIDTH and r.height >= MIN_ZONE_HEIGHT
    ]
    zones.sort(key=lambda z: z.area, reverse=True)
    return zones


def _compute_text_size(text: str, font_size: float, line_height: float, image_width: float, image_height: float) -> tuple[float, float]:
    lines = text.split("\n")
    longest_line = max((len(line) for line in lines), default=1)
    width_px = longest_line * font_size * 0.6
    height_px = len(lines) * font_size * line_height
    return round(width_px / image_width * 1000), round(height_px / image_height * 1000)


def assign_text_to_zones(
    suggestions: list[TextSuggestion],
    available_zones: list[SafeZone],
    image_width: float = 1000,
    image_height: float = 1000,
) -> list[dict]:
    if not available_zones:
        return [{"suggestion": s} for s in suggestions]

    zone_offsets: dict[str, float] = {z.label: z.top + MARGIN for z in available_zones}
    results: list[dict] = []

    for s in suggestions:
        font_size = (s.style or {}).get("font_size_normalized", 40)
        line_height = (s.style or {}).get("line_height", 1.2)
        text_w, text_h = _compute_text_size(s.part or "", font_size, line_height, image_width, image_height)

        target_zone = next((z for z in available_zones if z.label == s.preferred_zone), None)
        if not target_zone:
            target_zone = next(
                (z for z in available_zones if (zone_offsets.get(z.label, z.top + MARGIN) + text_h + MARGIN < z.top + z.height)),
                available_zones[0],
            )

        current_top = zone_offsets.get(target_zone.label, target_zone.top + MARGIN)
        placed_left = target_zone.left + MARGIN
        clamped_width = min(text_w, target_zone.width - MARGIN * 2)

        entry = {}
        if hasattr(s, '__dict__'):
            entry.update(vars(s))
        entry["position"] = BBox(
                top=max(target_zone.top + MARGIN, current_top),
                left=placed_left,
                width=max(50, clamped_width),
                height=max(20, text_h),
            )
        results.append(entry)

        zone_offsets[target_zone.label] = current_top + text_h + MARGIN

    return results
